drivers/hmis_csv_twenty_twenty_four/app/models/hmis_csv_twenty_twenty_four/importer/import_concern.rb

```diff
@@ -0,0 +1,297 @@
+      involved_warehouse_scope(
+        data_source_id: data_source_id,
+        project_ids: project_ids,
+        date_range: date_range,
+      ).with_deleted.
+        update_all(pending_date_deleted: pending_date_deleted)
+    end
+
+    def self.new_data(data_source_id:, project_ids:, date_range:, importer_log_id:)
+      existing_keys = involved_warehouse_scope(
+        data_source_id: data_source_id,
+        project_ids: project_ids,
+        date_range: date_range,
+      ).select(hud_key)
+      existing_keys = existing_keys.with_deleted if paranoid?
+
+      where(importer_log_id: importer_log_id).should_import.where.not(hud_key => existing_keys)
+    end
+
+    def self.existing_data(data_source_id:, project_ids:, date_range:)
+      existing_scope = involved_warehouse_scope(
+        data_source_id: data_source_id,
+        project_ids: project_ids,
+        date_range: date_range,
+      )
+      existing_scope = existing_scope.with_deleted if paranoid?
+      existing_scope
+    end
+
+    def self.incoming_data(importer_log_id:)
+      where(importer_log_id: importer_log_id).should_import
+    end
+
+    def self.existing_destination_data(data_source_id:, project_ids:, date_range:)
+      involved_warehouse_scope(
+        data_source_id: data_source_id,
+        project_ids: project_ids,
+        date_range: date_range,
+      ).with_deleted.
+        delete_pending
+    end
+
+    def self.pending_deletions(data_source_id:, project_ids:, date_range:)
+      involved_warehouse_scope(
+        data_source_id: data_source_id,
+        project_ids: project_ids,
+        date_range: date_range,
+      ).delete_pending
+    end
+
+    def as_destination_record
+      # For some odd reason calling build_destination_record sometimes does a find and update
+      # using new seems safer
+      klass = self.class.reflect_on_association(:destination_record).klass
+      record = klass.new(
+        attributes.slice(*self.class.create_columns),
+      )
+      record.source_hash = source_hash
+      # Note which record we're sending this from for error checking
+      record.source_id = id
+      # If we're creating a new record from a previously deleted
+      # record, make sure we bring it back to life
+      record.DateDeleted = self.DateDeleted if klass.paranoid?
+      record
+    end
+
+    def self.create_columns
+      (hmis_structure.keys + [:data_source_id]).map(&:to_s).freeze
+    end
+
+    def self.fix_date_columns(row)
+      date_columns.each do |col|
+        row[col] = fix_date_format(row[col])
+      end
+      row
+    end
+
+    def self.fix_time_columns(row)
+      time_columns.each do |col|
+        row[col] = fix_time_format(row[col])
+      end
+      row
+    end
+
+    # HMIS CSV FORMAT SPECIFICATIONS says under "Data Types"
+    # https://hudhdx.info/Resources/Vendors/HMIS%20CSV%20Specifications%20FY2020%20v1.8.pdf
```

Comment:
This seems a bit outdated now.  I assume the information hasn't changed, maybe we should just remove the link?